### PR TITLE
fix publishing

### PR DESCRIPTION
**Pull request by dryajov**
This is a WIP to fix some publishing issues, it duplicates some of the work in #244 so we might consider cherry picking either into this branch or into the #244 branch. I'm leaving it here for now as a reference ( @sinkingsugar  )

### CODE REVIEW

**Comment by sinkingsugar** [comment]
I like it.. I wanted to do this refactor but it would have taken me much more to figure it out.. don't worry about my further changes.. I just added some locks to heartbeat.. which worked 99% of the times.. still triggered nil peer tho

File: tests/pubsub/testfloodsub.nim [hunk 1]
```diff
@@ -60,7 +60,7 @@ suite "FloodSub":
       await nodes[1].subscribe("foobar", handler)
       await waitSub(nodes[0], nodes[1], "foobar")
 
-      await nodes[0].publish("foobar", "Hello!".toBytes())
+      discard await nodes[0].publish("foobar", "Hello!".toBytes())
```
Comment (arnetheduck) [inline tests/pubsub/testfloodsub.nim:63]: `check: await... == 1` (or whatever it should be?)

Comment (arnetheduck) [reply tests/pubsub/testfloodsub.nim:63]: `check: 1 == await ... ` maybe easier to read

Comment (sinkingsugar) [reply tests/pubsub/testfloodsub.nim:63]: that's from my branch btw 😄 , the thing is that this does not apply to flood, it will always be 0.

Comment (sinkingsugar) [reply tests/pubsub/testfloodsub.nim:63]: btw I find it useful (nbc side for example I'm using it) to have this int with gossip. but we can remove it once things stabilize if you guys don't think it's necessary.. it's a debug tool tho

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 2]
```diff
@@ -181,36 +206,46 @@ proc dropFanoutPeers(g: GossipSub) {.async.} =
...
-              result[id] = ControlMessage()
-            result[id].ihave.add(ihave)
+    for id in allPeers:
+      if result.len >= GossipSubD:
+        trace "got gossip peers", peers = result.len
+        break
 
-    libp2p_gossipsub_peers_per_topic_gossipsub
-      .set(g.gossipsub.getOrDefault(topic).len.int64, labelValues = [topic])
+      if allPeers.len == 0:
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:238]: allpeers can't be zero in here

**Pull request merged**